chainer/functions/connection/multihead_attention.py

```diff
@@ -0,0 +1,298 @@
+                    key_padding_mask.shape[0], key_padding_mask.shape[1]))
+
+    if add_zero_attn:
+        source_length += 1
+        k = concat.concat((
+            k,
+            generate_zeros(device, (len(k), 1) + k.shape[2:], dtype)
+        ))
+        v = concat.concat((
+            v,
+            generate_zeros(device, (len(v), 1) + v.shape[2:], dtype)
+        ))
+        if attn_mask is not None:
+            attn_mask = concat.cocnat((
+                attn_mask,
+                generate_zeros(device, (len(attn_mask), 1), dtype)
+            ))
+        if key_padding_mask is not None:
+            key_padding_mask = concat.concat(
+                (
+                    key_padding_mask,
+                    generate_zeros(
+                        device, (len(key_padding_mask), 1),
+                        key_padding_mask.dtype)
+                )
+            )
+    attn_output_weights = matmul.matmul(
+        q, transpose.transpose(k, (0, 2, 1)))
+    if (attn_output_weights.shape !=
+            (batch_size * n_head, target_length, source_length)):
+        raise ValueError('`attn_output_weights` is shaped wrongly')
+
+    if attn_mask is not None:
+        attn_mask = expand_dims.expand_dims(attn_mask, 0)
+        attn_output_weights += attn_mask
+
+    if key_padding_mask is not None:
+        attn_output_weights = reshape.reshape(
+            attn_output_weights,
+            (batch_size, n_head, target_length, source_length)
+        )
+        expanded_mask = expand_dims.expand_dims(
+            expand_dims.expand_dims(key_padding_mask, 1), 2)
+        attn_output_weights = where.where(
+            expanded_mask,
+            -xp.inf * generate_ones(
+                device, attn_output_weights.shape, dtype),
+            attn_output_weights
+        )
+        attn_output_weights = reshape.reshape(
+            attn_output_weights,
+            (batch_size * n_head, target_length, source_length)
+        )
+
+    attn_output_weights = softmax.softmax(attn_output_weights, axis=-1)
+    if attention_dropout > 0.0:
+        attn_output_weights = dropout.dropout(
+            attn_output_weights, attention_dropout)
+
+    attn_output = matmul.matmul(attn_output_weights, v)
+    attn_output = transpose.transpose(
+        attn_output,
+        (1, 0) + tuple(range(2, attn_output.ndim))
+    )
+    attn_output = reshape.reshape(
+        attn_output, (target_length, batch_size, embedding_size))
+    attn_output = linear.linear(
+        attn_output, proj_out_W, proj_out_b,
+        n_batch_axes=attn_output.ndim-1)
+    if post_dropout > 0.0:
+        attn_output = dropout.dropout(attn_output, post_dropout)
+
+    if return_weights:
+        attn_output_weights = reshape.reshape(
+            attn_output_weights,
+            (batch_size, n_head, target_length, source_length)
+        )
+        attn_output_weights = average.average(attn_output_weights, axis=1)
+    else:
+        attn_output_weights = None
```

Comment:
Could you motivate why it's better to return `None` instead of not retuning this at all? Just curious because the latter is more commonly in existing functions such as dropout and Gaussian sampling.

Reply:
I don't think it's better or not, but I wanted to make the return type consistent (`tuple`).